File: acme_checker.py
import dns.resolver

# Define a resolver instance globally to reuse for all queries
resolver = dns.resolver.Resolver()

def get_cname(domain):
    try:
        # Query for the CNAME record
        cname_answer = resolver.resolve('_acme-challenge.' + domain, 'CNAME')
        cname = cname_answer[0].target.to_text()
        return cname

    except dns.resolver.NoAnswer:
        return None
    except dns.resolver.NXDOMAIN:
        return None
    except dns.resolver.NoNameservers:
        return None

def main():
    # Improved error handling when the file is not found or inaccessible
    try:
        with open('acme_list.txt', 'r') as f:
            domains = f.read().splitlines()
    except FileNotFoundError:
        print("The file 'acme_list.txt' was not found.")
        return
    except IOError as e:
        print(f"An error occurred while trying to read the file: {e}")
        return

    for domain in domains:
        cname = get_cname(domain)
        if cname:
            print(f"CNAME for _acme-challenge.{domain} is {cname}")
        else:
            print(f"CNAME record missing for _acme-challenge.{domain}")

if __name__ == "__main__":
    main()

File: acme_checker.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"

	"github.com/miekg/dns"
)

func getCNAME(domain string) (string, error) {
	c := dns.Client{}
	m := dns.Msg{}
	m.SetQuestion(dns.Fqdn("_acme-challenge."+domain), dns.TypeCNAME)

	r, _, err := c.Exchange(&m, "8.8.8.8:53") // Using Google's public DNS server
	if err != nil {
		return "", err
	}
	if len(r.Answer) < 1 {
		return "CNAME record missing", nil
	}

	for _, ans := range r.Answer {
		if t, ok := ans.(*dns.CNAME); ok {
			return t.Target, nil
		}
	}
	return "CNAME record missing", nil
}

func main() {
	file, err := os.Open("acme_list.txt")
	if err != nil {
		log.Fatalf("Failed to open file: %s", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		domain := scanner.Text()
		cname, err := getCNAME(domain)
		if err != nil {
			fmt.Printf("Error querying CNAME for %s: %s\n", domain, err)
			continue
		}
		fmt.Printf("CNAME for _acme-challenge.%s is %s\n", domain, cname)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading file: %s", err)
	}
}

